### PR TITLE
Add exception filters to make crash dump analysis easier

### PR DESCRIPTION
**Pull request by tmat**

### CODE REVIEW

**Comment by tmat** [comment]
@jasonmalinowski @gafter @mavasani

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/CompilationWithAnalyzers.cs [hunk 1]
```diff
@@ -457,146 +478,174 @@ public Task<ImmutableArray<Diagnostic>> GetAllDiagnosticsAsync()
...
 
-                cancellationToken.ThrowIfCancellationRequested();
+                // Populate the events cache from the generated compilation events.
+                await PopulateEventsCacheAsync(cancellationToken).ConfigureAwait(false);
 
-                // Track if this task was suspended by another tree diagnostics request for the same tree.
-                // If so, we wait for the high priority requests to complete before restarting analysis.
-                bool suspendend;
-                do
+                try
```
Comment (gafter) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/CompilationWithAnalyzers.cs]: I recommend fusing these try statements. You probably don't want to do the cleanup in the first finally before capturing the failure.

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/CompilationWithAnalyzers.cs [hunk 2]
```diff
@@ -457,146 +478,174 @@ public Task<ImmutableArray<Diagnostic>> GetAllDiagnosticsAsync()
...
+                                                {
+                                                    // Get event queue with pending events to analyze.
+                                                    eventQueue = getEventQueue();
+
+                                                    // Execute analyzer driver on the given analysis scope with the given event queue.
+                                                    await ComputeAnalyzerDiagnosticsCoreAsync(driver, eventQueue, analysisScope, cancellationToken: linkedCts.Token).ConfigureAwait(false);
+                                                }
+                                                finally
+                                                {
+                                                    FreeEventQueue(eventQueue);
```
Comment (gafter) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/CompilationWithAnalyzers.cs]: Is it important to free the event queue before reporting the fatal error? I recommend fusing these try statements too.

**Comment by gafter** [comment]
LGTM. I did notice a couple of places where you have a try-finally nested within a try-catch, and the former does some cleanup while the latter logs a fatal exception. In these cases I recommend fusing the two into a single try-catch-finally, so the fatal exception is logged before the cleanup.

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/CompilationWithAnalyzers.cs (hunk 2 above)
Comment (tmat) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/CompilationWithAnalyzers.cs]: The pattern we use for the filters is to wrap the entire method body. The goal of the filter is to intercept an exception coming from any statement of the method body before it reaches the synthesized catch-all clause that the compiler generates around the entire method body.

The final IL looks as follows:
```
try
{
  try { original body } filter { crash }  
}
catch (Exception e)      // compiler generated
{                          
   task.Exception = e;
}
```

Comment (tmat) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/CompilationWithAnalyzers.cs]: Note that the filter condition is executed before any finally blocks run and before the stack is unwound.

Comment (gafter) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/CompilationWithAnalyzers.cs]: Right, I forgot that. Never mind.

**Comment by tmat** [comment]
@dotnet-bot retest this please

**Pull request merged**